File: Frontend/trackr-administrador/src/app/shared/http/examen/contenido-examen.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { ContenidoExamen } from '@models/examen/contenido-examen';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root',
})
export class ContenidoExamenService {
  private dataUrl = 'contenidoExamen/';

  constructor(public http: HttpClient) {}

  public consultarTodosParaSelector(): Observable<ContenidoExamen[]> {
    return this.http.get<ContenidoExamen[]>(this.dataUrl + 'consultarTodosParaSelector');
  }

  public consultar(idContenidoExamen: number): Observable<ContenidoExamen> {
    return this.http.get<ContenidoExamen>(this.dataUrl + `consultar/${idContenidoExamen}`);
  }

  public consultarGeneral(idTipoExamen: number): Observable<ContenidoExamen[]> {
    return this.http.get<ContenidoExamen[]>(this.dataUrl + `consultarGeneral/${idTipoExamen}`);
  }

  public agregar(contenidoExamen: ContenidoExamen): Observable<number> {
    return this.http.post<number>(this.dataUrl + 'agregar', contenidoExamen);
  }

  public editar(contenidoExamen: ContenidoExamen): Observable<void> {
    return this.http.put<void>(this.dataUrl + 'editar', contenidoExamen);
  }

  public eliminar(idContenidoExamen: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idContenidoExamen}`);
  }
}

File: Backend/TrackrAPI/TrackrAPI/Services/GestionExamen/ContenidoExamenService.cs
using TrackrAPI.Dtos.GestionExamen;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.GestionExamen;

namespace TrackrAPI.Services.GestionExamen;

public class ContenidoExamenService
{
    private readonly IContenidoExamenRepository _contenidoExamenRepository;
    private readonly ContenidoExamenValidatorService _contenidoExamenValidatorService;
    private readonly TipoExamenService _tipoExamenService;

    public ContenidoExamenService(
        IContenidoExamenRepository contenidoExamenRepository,
        ContenidoExamenValidatorService contenidoExamenValidatorService,
        TipoExamenService tipoExamenService)
    {
        _contenidoExamenRepository = contenidoExamenRepository;
        _contenidoExamenValidatorService = contenidoExamenValidatorService;
        _tipoExamenService = tipoExamenService;
    }

    public ContenidoExamenDto? Consultar(int idContenidoExamen)
    {
        return _contenidoExamenRepository.Consultar(idContenidoExamen);
    }

    public IEnumerable<ContenidoExamenGridDto> ConsultarGeneral(int idTipoExamen)
    {
        return _contenidoExamenRepository.ConsultarGeneral(idTipoExamen);
    }

    public IEnumerable<ContenidoExamenGridDto> ConsultarTodosParaSelector()
    {
        return _contenidoExamenRepository.ConsultarTodosParaSelector();
    }

    public int Agregar(ContenidoExamen contenidoExamen)
    {
        _contenidoExamenValidatorService.ValidarAgregar(contenidoExamen);
        _contenidoExamenRepository.Agregar(contenidoExamen);
        _tipoExamenService.CalcularReactivos(contenidoExamen.IdTipoExamen);
        return contenidoExamen.IdContenidoExamen;
    }

    public void Editar(ContenidoExamen contenidoExamen)
    {
        _contenidoExamenValidatorService.ValidarEditar(contenidoExamen);
        _contenidoExamenRepository.Editar(contenidoExamen);
        _tipoExamenService.CalcularReactivos(contenidoExamen.IdTipoExamen);
    }

    public void Eliminar(int idContenidoExamen)
    {
        ContenidoExamenDto? contenidoExamen = _contenidoExamenRepository.Consultar(idContenidoExamen);
        _contenidoExamenValidatorService.ValidarEliminar(idContenidoExamen);

        if (contenidoExamen != null)
        {
            ContenidoExamen contenidoExamenMod = new()
            {
                IdContenidoExamen = contenidoExamen.IdContenidoExamen,
                IdTipoExamen = contenidoExamen.IdTipoExamen,
                IdAsignatura = contenidoExamen.IdAsignatura,
                IdNivelExamen = contenidoExamen.IdNivelExamen,
                Clave = contenidoExamen.Clave,
                TotalPreguntas = contenidoExamen.TotalPreguntas,
                Duracion = contenidoExamen.Duracion,
                FechaAlta = contenidoExamen.FechaAlta,
                Estatus = false
            };

            _contenidoExamenRepository.Editar(contenidoExamenMod);
            _tipoExamenService.CalcularReactivos(contenidoExamen.IdTipoExamen);
        }
    }
}
